inventory_management/inventory_management/doctype/stock_entry/stock_entry.py

```diff
@@ -0,0 +1,215 @@
+                )
+
+    def before_insert(self):
+        self.stock_entry_items = club_similar_item_rows(self.stock_entry_items)
+
+    def on_submit(self):
+
+        for item_row in self.stock_entry_items:
+            receipt_valuation_rate = get_valuation_rate(
+                item_row.item, item_row.rate, item_row.qty
+            )
+            consume_valuation_rate = get_valuation_rate(
+                item_row.item, item_row.rate, -item_row.qty
+            )
+
+            if self.type == "Receive":
+                create_sle(
+                    item_row.target_warehouse,
+                    item_row.qty,
+                    item_row.item,
+                    receipt_valuation_rate,
+                    self.date,
+                    self.time
+                )
+
+            elif self.type == "Consume":
+                create_sle(
+                    item_row.source_warehouse,
+                    -item_row.qty,
+                    item_row.item,
+                    consume_valuation_rate,
+                    self.date,
+                    self.time
+                )
+
+            else:
+                create_sle(
+                    item_row.target_warehouse,
+                    item_row.qty,
+                    item_row.item,
+                    receipt_valuation_rate,
+                    self.date,
+                    self.time
+                )
+                create_sle(
+                    item_row.source_warehouse,
+                    -item_row.qty,
+                    item_row.item,
+                    consume_valuation_rate,
+                    self.date,
+                    self.time
+                )
+
+    def on_cancel(self):
+        for item in self.stock_entry_items:
+            receipt_valuation_rate = get_valuation_rate(item.item, item.rate, item.qty)
+            consume_valuation_rate = get_valuation_rate(item.item, item.rate, -item.qty)
+
+            if self.type == "Receive":
+                create_sle(
+                    item.target_warehouse, -item.qty, item.item, consume_valuation_rate,self.date,
+                    self.time
+                )
+
+            elif self.type == "Consume":
+                create_sle(
+                    item.source_warehouse, item.qty, item.item, receipt_valuation_rate,self.date,
+                    self.time
+                )
+
+            else:
+                create_sle(
+                    item.target_warehouse, -item.qty, item.item, consume_valuation_rate,self.date,
+                    self.time
+                )
+                create_sle(
+                    item.source_warehouse, item.qty, item.item, receipt_valuation_rate,self.date,
+                    self.time
+                )
+
+
+def club_similar_item_rows(stock_entry_items):
+    dict_items = {}
+    for item_row in stock_entry_items:
+        if not (item_row.item in dict_items):
+            dict_items[item_row.item] = item_row
```

Comment:
We need to club same item-warehouse pair not just similar items

Item A - Warehouse A
Item A - Warehouse B

these 2 shouldn't be clubbed. Only same item-warehouse should be clubbed